neptunecontrib/hpo/utils.py

```diff
@@ -0,0 +1,187 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+import random
+import string
+import subprocess
+
+from neptunelib.session import Session
+from scipy.optimize import OptimizeResult
+import skopt
+
+
+def make_objective(param_set, command, metric_channel_name, project_name, tag='trash'):
+    """Converts `neptune run` script to a python function that returns a score.
+
+    Helper function that wraps your train_evaluate_model.py script into a function
+    that takes hyper parameters as input and returns the score on a specified metric.
+
+    Args:
+        param_set(dict): Dictionary of parameter name parameter value. The objective
+            function is evaluated on those parameters.
+        command(list):bash neptune command represented as a list. It is assumed
+            that the .py script path is a last element.
+            E.g. ['neptune run --config neptune.yaml', 'main.py']
+        metric_channel_name(str): name of the property where the evaluation
+            result is stored. It is crucial that the single run result is stored
+            as property.E.g. ctx.properties[metric_channel_name] = 0.8.
+        project_name(str): NAME_SPACE/PROJECT_NAME. E.g. 'neptune-ml/neptune-examples'
+        tag(str): Tag that should be added to the single run experiment.
+            It is useful to add this tag so that you can group or get rid of the single run
+            experiments in your project. Default is 'trash'.
+
+    Returns:
+        float: Score that was reported in the `ctx.properties[metric_channel_name]`
+        of the run of train_evaluate_model.py script.
+
+
+    Examples:
+        Prepare the space of hyperparameters to search over.
+
+        >>> space = [skopt.space.Integer(10, 60, name='num_leaves'),
+                     skopt.space.Integer(2, 30, name='max_depth'),
+                     skopt.space.Real(0.1, 0.9, name='feature_fraction', prior='uniform')]
+
+        Define the objective.
+
+        >>> @skopt.utils.use_named_args(space)
+            def objective(**params):
+                return hp_utils.make_objective(params,
+                                               command=['neptune run --config neptune.yaml',
+                                               'train_evaluate.py'],
+                                               metric_channel_name='valid_accuracy',
+                                               project_name='neptune-ml/neptune-examples')
+
+        Run skopt training.
+
+        >>> results = skopt.forest_minimize(objective, space,
+                                base_estimator='ET', n_calls=100, n_random_starts=10)
+
+    """
+    command, exp_id_tag = _create_command_and_id_tag(command, param_set, tag)
+
+    subprocess.call(command, shell=True)
+
+    score = _get_score(exp_id_tag, metric_channel_name, project_name)
```

Comment:
@pitercl this module makes it possible to run all the bayesian hyper parameter searches on any neptune script.

It heavily relies on the neptunelib api since it fetches the .property of the tagged experiment.
What it means is that sometimes, when the server is unavailable it fails to fetch the score of the experiment.

How would you handle that? 
Would some amount of retries be ok in your opinion?

Reply:
I could maybe decorate `_get_score` function with something like this:
https://www.saltycrane.com/blog/2009/11/trying-out-retry-decorator-python/

Reply:
If this is critical and blows up user's experimentation process in case of request failure, then yes, retries are in order.

Reply:
Done.